File: aoc23/2/main.py

f = open("2/test.txt")
f = open("2/input.txt")

MAX_RED = 12
MAX_GREEN = 13
MAX_BLUE = 14

class ValidationError(Exception):
    pass

def validate_draws(draws, check_max=True):
    max_red = 0
    max_green = 0
    max_blue = 0
    for draw in draws.split(";"):
        draw = draw.strip()
        # print(draw)
        for part in draw.split(","):
            part = part.strip()
            # print(part)
            number, color = part.split(" ")
            number = int(number)
            if color == "red":
                max_red = max(max_red, number)
                if check_max and number > MAX_RED:
                    print(f"invalid game id: {game_id}, because of RED")
                    raise ValidationError
            elif color == "green":
                max_green = max(max_green, number)
                if check_max and number > MAX_GREEN:
                    print(f"invalid game id: {game_id}, because of GREEN")
                    raise ValidationError
            elif color == "blue":
                max_blue = max(max_blue, number)
                if check_max and number > MAX_BLUE:
                    print(f"invalid game id: {game_id}, because of BLUE")
                    raise ValidationError
    return max_red, max_green, max_blue


valid_game_ids = []
valid_game_powers = []
for l in f.readlines():
    l = l.strip()
    print(l)
    game_id, draws = l.split(":")
    game_id = game_id.removeprefix("Game ")
    game_id = int(game_id)
    try:
        validate_draws(draws)
        valid_game_ids.append(game_id)
    except ValidationError:
        pass
    r,g,b = validate_draws(draws, check_max=False)
    valid_game_powers.append(r*g*b)

print(valid_game_ids)
print(sum(valid_game_ids))
print(valid_game_powers)
print(sum(valid_game_powers))

File: aoc23/2/src/main.rs
use itertools::Itertools;
use std::env;
use std::fs;

fn main() {
    let mut current_dir = env::current_dir().unwrap().to_str().unwrap().to_owned();
    if !current_dir.ends_with("/2") {
        current_dir += "/2"
    }
    let contents = fs::read_to_string(current_dir + "/input.txt").expect("couldn't read file");
    let lines = contents.lines();

    // println!("{:?}", lines);
    let result = lines
        .into_iter()
        .map(|l| {
            println!("{:?}", l);
            // l contains: Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
            // add max of each color to numbers
            l.split_once(':')
                .unwrap()
                .1
                .split(';')
                .map(|part| {
                    part.split(',').fold([0, 0, 0], |mut set, c| {
                        match c.trim().split_once(' ') {
                            Some((num, "red")) => set[0] += num.parse::<usize>().unwrap(),
                            Some((num, "blue")) => set[1] += num.parse::<usize>().unwrap(),
                            Some((num, "green")) => set[2] += num.parse::<usize>().unwrap(),
                            _ => {
                                println!("{:?}", c);
                                unreachable!("bad input")
                            }
                        }
                        set
                    })
                })
                .collect_vec()
        })
        .map(|game| {
            println!("{:?}", game);
            game.iter()
                .fold([0, 0, 0], |set, part| {
                    [
                        set[0].max(part[0]),
                        set[1].max(part[1]),
                        set[2].max(part[2]),
                    ]
                })
                .iter()
                .product::<usize>()
        })
        .collect_vec();

    println!("{:?}", result);
    println!("{:?}", result.iter().sum::<usize>());
}
